### ISSUE TITLE
False-positives in `no-fn-reference-in-iterator` rule

### ISSUE THREAD

**Issue by modosc** [sindresorhus/eslint-plugin-unicorn#121]
[<img alt="Issuehunt badges" src="https://issuehunt.io/r/sindresorhus/eslint-plugin-unicorn/issues/121/badge.svg" />](https://issuehunt.io/r/sindresorhus/eslint-plugin-unicorn/issues/121)

the `unicorn/no-fn-reference-in-iterator` matched all of these pieces of (valid) code and the fix applied broke every one. not sure the best way forward for this..

https://reactjs.org/docs/react-api.html#reactchildrenforeach
```
React.Children.forEach(children, (p) => {
})
```

https://github.com/airbnb/enzyme/blob/master/docs/api/ReactWrapper/find.md
```
const wrapper = shallow(<Foo />)
wrapper.find(Panel).first().childAt(1)
```

https://caolan.github.io/async/docs.html#filter
```
    async.filter(
      Object.keys(manifest),
      (f, callback) => {
        callback(null, f.match(/\.js$/))
      },
      (err, results) => {
        async.each(results, (f, callback) => {
```

https://lodash.com/docs/4.17.4#forEach
```
      _.forEach(api, (op, m) => {
```

https://github.com/react-bootstrap/react-bootstrap/blob/master/src/utils/ValidComponentChildren.js

```
import ValidComponentChildren from 'react-bootstrap/lib/utils/ValidComponentChildren'
ValidComponentChildren.map(children, child => (
```

> There is a $40.00 open bounty on this issue. Add more on [Issuehunt](https://issuehunt.io/r/sindresorhus/eslint-plugin-unicorn/issues/121).
> - Checkout the [Issuehunt explorer](https://issuehunt.io/r/sindresorhus/eslint-plugin-unicorn/) to discover more funded issues.
> - Need some help from other developers? [Add your repositories](https://issuehunt.io/r/new) on Issuehunt to raise funds.

**Comment by AdriVanHoudt** [comment]
Getting the same thing. 
The auto fix results in 
```diff
-    return Async.map(stages, (stage, callback) => {
+    return Async.map((x) => stages(x), (stage, callback) => {
```

**Comment by sindresorhus** [comment]
Unfortunately, there's really no way to know the actual type of the variable. So this rule is destined to have some false positives. We should more clearly mention this in the docs.

Like we did with https://github.com/sindresorhus/eslint-plugin-unicorn/commit/53dc7c11c187e656d2e832e2b02ce830bf417934, we could add some overrides for known globals, like `React.Children`, `async`, `_`. We could also only report when there's just one argument to the method.

**Comment by zimme** [comment]
Should this really be in the recommended config if it's prone to false positives?

**Comment by wmertens** [comment]
The "one argument" heuristic seems like a great idea…

**Comment by johnwu93** [comment]
For dealing with these false-positives, I can implement what @sindresorhus suggested. I can add configuration rules to this rule so that the user has the ability to exclude identifiers from this rule. Also, I can add a configuration rule where we can report methods with one argument, which will ignore the errors for `React.Children` and `Async `. How does that sound?

**Comment by sindresorhus** [comment]
> I can add configuration rules to this rule so that the user has the ability to exclude identifiers from this rule.

👍

> Also, I can add a configuration rule where we can report methods with one argument

👍, but it should not be configurable. We should always ignore methods with more than 1 argument.

**Comment by MrHen** [comment]
FWIW, [lodash/prefer-lodash-method](https://github.com/wix/eslint-plugin-lodash/blob/master/docs/rules/prefer-lodash-method.md#rule-details) had to solve this problem too and they chose to do it by adding `ignoreMethods` and `ignoreObjects` configs to the rule.

**Comment by sindresorhus** [comment]
Reopening as this is not completely fixed.

**Comment by sindresorhus** [comment]
Someone still needs to handle https://github.com/sindresorhus/eslint-plugin-unicorn/pull/172#issuecomment-383017266.

**Comment by IssuehuntBot** [comment]
@issuehunt has funded $40.00 to this issue.

---
- Submit pull request via [IssueHunt](https://issuehunt.io/repos/55832243/issues/121) to receive this reward.
- Want to contribute? Chip in to this issue via [IssueHunt](https://issuehunt.io/repos/55832243/issues/121).
- Checkout the [IssueHunt Issue Explorer](https://issuehunt.io/issues) to see more funded issues.
- Need help from developers? [Add your repository](https://issuehunt.io/r/new) on IssueHunt to raise funds.
